Added a per thread buffer and related tests for OMR socket API

caohaley opened the pull request:
Created 4 commits for this PR:

-  `Added Omrsock Related Per Thread Buffer functions` contains changes related to the per thread buffer and its functionalities.
- `Added startup and shutdown functions to omrsock API` contains changes related to OMR socket API startup and shutdown, which utilizes the ptb functions in the previous commit to startup and shutdown the per thread buffer.
- `Added Unix implementation for omrsock_getaddrinfo_create_hints` contains implementation of the OMR socket API function, that uses the ptb function to store information. It also contains a few new error codes related to OMR socket API.
- `Added Implementations to omrsockTest.cpp to test omrsock PTB` contains tests that are still being work on for the per thread buffer.

caohaley commented:
I ran into a few problems with this PR.

First, I need to include `omrsock.h` in `omrsockTest.cpp` to access elements of the struct `OMRAddrInfoNode`. However, I'm not sure how to include the file without specifying a path to it.

Also, when we run `omrsockTest.cpp`, for windows systems, it won't be able to find a `omrsock.h` because I haven't implemented for windows systems yet. It also will only read `common/omrsock.c` where the implementations will only return errors. It is possible to only run the tests on Unix systems? Should I just wrap the test code in `per_thread_buffer_functionality` test with #if defined (Unix) or something similar to that?

babsingh commented:
> Need to include omrsock.h in omrsockTest.cpp. How to include the file without specifying a path to it?

You will have to include a path, and use `#if defined(OMR_OS_WINDOWS)` to distinguish between Windows and Unix paths.

> Second query (Windows unimplemented).

When you start implementing for Unix, you will need to include stubs for Windows. Windows won't pick up from `common/omrsock.c`. @rwy0717 Is this correct?

You can disable the test on Windows by updating the makefile; exclude the `omrsockTest` object using `ifeq (win,$(OMR_HOST_OS))`.

Refer to the code and make files in `omr/fvtest/porttest`.

caohaley commented:
@babsingh Is it possible to move contents of `unix_include/omrsock.h` into `omrportsock.h` so that the users of the socket API can just include `omrportsock.h`?

When I include the path to `omrsock.h`, there is the problem of `conflicting declaration` between `omrsock.h` 's definition vs. `omrportsock.h`'s forward declaration.

```/root/omr/include_core/../port/unix_include/omrsock.h:35:17: error: conflicting declaration 'typedef int32_t OMRSocket'
 typedef int32_t OMRSocket;
                 ^
In file included from /root/omr/include_core/omrport.h:46:0,
                 from /root/omr/fvtest/porttest/omrsockTest.cpp:22:
/root/omr/include_core/omrportsock.h:33:16: error: 'struct OMRSocket' has a previous declaration as 'struct OMRSocket'
 typedef struct OMRSocket *omrsock_socket_t;
                ^
```
I think it may be because the compiler is confused between `struct Type` for forward declaration vs `Type` in typedef. It may think that they are 2 different definitions for the same Type name.

babsingh commented:
You are likely to cause redefinition errors as before if you expose struct definitions to the OMR consumer via `omrportsock.h`.

You are getting the compile error because `OMRSocket` can't be  an `int32_t` and `struct` at the same time.

It needs to be 
```
typedef struct OMRSocket {
     int32_t socket;
} OMRSocket;
```

Let me know if this works.

caohaley commented:
Hi Babneet, sorry for the late reply. That method definitely works after I added a struct for `OMRSocket` and `OMRSockAddrStorage`. 

However, I'm having trouble picturing how that will work for the users because I think the users will need to declare and allocate space for structures like `OMRSocket`,  `OMRSockAddrStorage`, and `OMRAddrInfoNode` although they won't directly access the elements inside. They will just pass a pointer to these structures, into the omrsock functions, to perform the operations they need. 

In `j9sock.c`, I think that's how they used the structures: https://github.com/eclipse/openj9/blob/a3a3664e6db6ee632a176c056fe0845bea322a29/runtime/tests/port/j9sockTest.c#L1400

babsingh commented:
@caohaley Can you elaborate on your issue with a simple use-case/example?

caohaley commented:
@babsingh  Basically, I'm still a bit confused on how the users will use our program. For example, in `j9sockTest.c`, the user allocates struct and pass in the pointer to the struct into the socket API, for example:
```
j9addrinfo_struct res;
j9addrinfo_t hints;
int32_t length;

j9sock_getaddrinfo_create_hints(&hints, ...);
/* Now hints is pointing to allocated memory in per thread buffer */

/* Pass in pointer to res */
j9sock_getaddrinfo("localhost", hints, &res);

/* To traverse the res, we use j9getaddrinfo_length, j9sock_getaddrinfo_family funtions */
j9sock_getaddrinfo_length(&res, &length);
```
I think the users may need to access the structs to allocate them locally, even though they won't need to access the elements inside.

caohaley commented:
However, I was talking to @rwy0717 and Robert brought up that in `j9sock.c`, the API allocates j9socket_struct for the users in `j9sock_socket` and frees the memory in `j9sock_close`. 

(Allocation in `j9sock_socket`: https://github.com/eclipse/openj9/blob/1dd14da6ca1351c8ab42f40b0c7e324c39ab102e/runtime/port/unix/j9sock.c#L3554.)

babsingh commented:
@rwy0717 is correct. Similarly, you can hide allocations for other structs in create/open/init methods, and de-allocate them during close/free/destroy methods.

caohaley commented:
I see, we could do that. But how would we keep track of what was allocated? For example, in `j9close` function, you would pass in the `j9socket_t` to be freed.

babsingh commented:
A user can do three things before invoking the omrsock API:
```
        //static allocation
	omrsockaddr_struct local_addrStruct;
	omrsockaddr_t local_addr = &local_addrStruct;

        //dynamic allocation
	omrsockaddr_t local_addr = malloc(...);

        //no allocation
	omrsockaddr_t local_addr = NULL;
```

Your API should handle all three scenarios. If your API does not handle all three scenarios, then you should clearly state the usage of your API in the description with examples.

> how would we keep track of what was allocated?

You can use `omrsocket_t` as a reference and internally cache the data structs associated with a `omrsocket_t` in a hash table (or in a similar <key, value> struct). During `create/open/init`, you create the `omrsocket_t` (key) and add an entry into the hash table which will contain all the structs associated with the socket. During `close/free/destroy`, you will query the hash table for the entry and free the structs associated with the `omrsocket_t`. You can also use an array instead of a hash table, and store the array index within `omrsocket_t`.

caohaley commented:
I see that makes a lot of sense, thank you Babneet! I will look into it more.

caohaley commented:
@babsingh Would it be helpful if we made the definition of the OMRSocket struct:
```
typedef OMRSocket{
    int socket;
    int curMemoryIndex;
    void * memoryManagement[30];
} OMRSocket;
```
When we allocate the OMRSocket struct for the user, we set curMemoryIndex to 0, and initialize all elements of memoryManagement to NULL.

When we need to allocate memory we copy the pointer to the block of memory we allocate, to the memoryManagement array and increment curMemoryIndex.

At the very end, we free the pointers to the memory one by one.

babsingh commented:
@caohaley What do you want to store for every `OMRSocket`? `void * memoryManagement[30]` allocates memory for 30 void pointers. What do you plan to store in them?

caohaley commented:
I was thinking we could store the pointers to all the structs that we allocate for the user. I used a maximum of 30 allocations per socket_t that for the user, as an example.

Or we could use the hash table and it would look more clean. However, it may require more lines of programming.

caohaley commented:
@babsingh For the static and dynamic allocation method, will we still have the user include both `omrsock.h` as well as `omrportsock.h`, where omrsock.h defines the structs and omrportsock.h defining the pointers? It just seems to me a little inconvenient for the users to find the exact path to `omrsock.h` when including the header, where we could directly include the definition of the structs in `omrportsock.h`.

> ```
> //static allocation
> omrsockaddr_struct local_addrStruct;
> omrsockaddr_t local_addr = &local_addrStruct;
> 
> //dynamic allocation
> omrsockaddr_t local_addr = malloc(...);
> ```

babsingh commented:
> For the static and dynamic allocation method

You can just do the allocation for them in your API, and mention usage of the API with examples in the description. Then, users won't have to worry about allocations and including header files.

> We could store the pointers to all the structs that we allocate for the user

You need to identify all the structs per socket. Create a struct like:
```
typedef OMRSocketMeta{
    struct_1;
    struct_2;
    struct_3;
    ...;
} OMRSocketMeta;

typedef OMRSocket{
    int socket;
    struct OMRSocketMeta* meta;
} OMRSocket;
```
Allocate/initialize meta at the start, and de-allocate meta at the end. If there are global structures (shared by the entire socket library), then store them in `J9PortLibrary`; you can allocate and de-allocate them in the startup/shutdown methods.

caohaley commented:
@babsingh  @rwy0717 I committed the changes regarding creating structs in `omrportsocktypes.h`. Please let me know if this is what you had in mind. I tried to only forward declare the structs in `omrportsocktypes.h` and putting everything else into `common/omrsock.h` so that the users can make objects of the structs but can't access elements within. But I found out that users would not be able to make objects of the struct if it is incomplete type, or if it has members that are forward declared. Therefore, I put the member elements inside the same file. This way, the users can still access too many things so this is probably not the best way. Do you have any other proposals on how to make it work?

Link to Code: https://github.com/eclipse/omr/pull/4555/commits/55450a92da9fbf73aee8811c9d109f4a56905ea0

The users would be able to create the structs using:
```
#include "omrportsocktypes.h"
OMRSockAddrStorage_struct testAddrStorageStruct;
OMRSocket_struct testSocketStruct;
OMRAddrInfoNode testAddrInfoNode;
```
As shown in the test (fvtest/porttest/omrsockTest.cpp):
https://github.com/eclipse/omr/pull/4555/commits/0de446db47e9c7969d00cc15dae468d1b446f718#

babsingh commented:
@caohaley You have header files misnamed in your previous comment. You have things in the correct place: `omrportsock.h` has the forward declarations and `omrportsocktypes.h` has the data-structure definitions.

rwy0717 commented on fvtest/porttest/omrsockTest.cpp:
```diff
@@ -72,7 +73,24 @@ connect_client_to_server(struct OMRPortLibrary *portLibrary, const char *addrStr
+    EXPECT_NE(OMRPORTLIB->sock_sendto, nullptr);
+    EXPECT_NE(OMRPORTLIB->sock_recv, nullptr);
+    EXPECT_NE(OMRPORTLIB->sock_recvfrom, nullptr);
+    EXPECT_NE(OMRPORTLIB->sock_close, nullptr);
```
Indentation should be tabs, you've got 4 spaces.

rwy0717 commented on fvtest/porttest/omrsockTest.cpp:
```diff
@@ -83,7 +101,21 @@ TEST(PortSockTest, library_function_pointers_not_null)
+
+    OMRSockAddrStorage_struct testAddrStorageStruct;
+    OMRSocket_struct testSocketStruct;
+    OMRAddrInfoNode testAddrInfoNode;
```
Indentation should be tabs.

rwy0717 commented on include_core/omrportsocktypes.h:
```diff
@@ -0,0 +1,86 @@
+
+typedef struct OMRSockAddrStorage_struct{
+	OMRSockAddrStorage addrStorage;
+}OMRSockAddrStorage_struct;
```
Formatting is off, needs spacing around "{" and "}".

rwy0717 replied:
I think the naming we're using here is a little confusing. The distinction between `OMRSockAddrStorage_struct` and `OMRSockAddrStorage` isn't very clear. It would be better to mark the types as either OS or OMR. Here's what I mean (but you may want to choose slightly different names):

```c++

/* Common data types required for using the omr socket API */

#if defined(OMR_OS_WINDOWS)
typedef SOCKET omr_os_socket;
typedef struct sockaddr_storage omr_os_sockaddr_storage;  /* For IPv4 or IPv6 addresses */
typedef struct addrinfoW omr_os_addrinfo;  /* addrinfo structure – Unicode, for IPv4 or IPv6 */
#else /* defined(OMR_OS_WINDOWS) */
typedef int32_t omr_os_socket;
typedef struct sockaddr_storage omr_os_sockaddr_storage;  /* For IPv4 or IPv6 addresses */
typedef struct addrinfo omr_os_addrinfo;  /* addrinfo structure for IPv4 or IPv6*/
#endif /* defined(OMR_OS_WINDOWS) */

typedef struct OMRSockAddrStorage {
    omr_os_sockaddr_storage data;
} OMRSockAddrStorage;
```

caohaley replied:
Sounds good! It looks a lot cleaner this way.

rwy0717 commented on include_core/omrportsocktypes.h:
```diff
@@ -0,0 +1,86 @@
+
+typedef struct OMRSocket_struct{
+	OMRSocket socket;
+}OMRSocket_struct;
```
Same formatting issue here.

rwy0717 replied:
Similar to above, I'm not a fan of the naming. I would prefer something like:
```c++
typedef struct OMRSocket {
    omr_os_socket data;
} OMRSocket;

rwy0717 commented on include_core/omrportsocktypes.h:
```diff
@@ -0,0 +1,86 @@
+	 * Pointer to the first addrinfo node in listed list. 
+	 */
+	OMRAddrInfo *addrInfo;
+	/* Number of addrinfo nodes in linked list */
```
This comment should be a doxygen comment:
```c++
/**
 * Number of addrinfo nodes in linked list.
 */
```

rwy0717 commented on include_core/omrportsocktypes.h:
```diff
@@ -0,0 +1,86 @@
+	OMRSocket socket;
+}OMRSocket_struct;
+
+/* Filled in using @ref omr_getaddrinfo. */
```
This should be a doxygen comment. You should add a description of what the node is, not just how it's used.
```c++
/**
 * A node in a linked-list of addrinfo. Filled in using @ref omr_getaddrinfo.
 */
```

rwy0717 commented on include_core/omrportsocktypes.h:
```diff
@@ -0,0 +1,86 @@
+typedef struct OMRAddrInfoNode {
+	/* Defined differently depending on the operating system.
+	 * Pointer to the first addrinfo node in listed list. 
+	 */
```
This should be a doxygen comment:
```c++
	/**
	 * Pointer to the first addrinfo node in listed list. Defined differently depending on the operating system.
	 */
```

rwy0717 commented:
I see some conditionals and ifdefs related to disabling socket support on windows. We should only be testing the feature flag `OMR_PORT_SOCKET_SUPPORT` in C/C++.

If the feature isn't available on windows, that should be a check in cmake. The user will need to know that sockets can't be enabled. In general, if a user enables an unsupported feature that should be a hard error in cmake.

You can write the test like this in cmake:
```
omr_assert(
    TEST OMR_PORT_SOCKET_SUPPORT AND NOT OMR_HOST_OS STREQUAL "win"
    MESSAGE "Socket functionality is not supported on windows"
)
```

This also means that we will have to disable socket support in windows CI builds.

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -0,0 +1,435 @@
+		return rc;
+	}
+    
+#define ptbHints (ptBuffer->addrInfoHints).addrInfo
```
Using macros like this is bad style. Just access the field directly.

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -0,0 +1,435 @@
+	addrInfoDataP->ai_protocol = protocol;
+#undef ptbHints
+
+	*hints = &(ptBuffer->addrInfoHints);
```
```suggestion
	*hints = &ptBuffer->addrInfoHints;
```

caohaley commented on port/common/omrsockptb.c:
```diff
@@ -0,0 +1,139 @@
+omrsock_ptb_get(struct OMRPortLibrary *portLibrary)
+{
+    omrthread_t self = omrthread_self();
+	OMRSocketPTB *ptBuffer = omrthread_tls_get(self, portLibrary->portGlobals->socketTlsKey);
```
Yes, it should have one.

caohaley commented:
@rwy0717 Sounds good! How do I disable socket support in CI builds by platform? Right now, I've enabled it in `/root/omr/cmake/caches/Travis.cmake` and `/root/omr/cmake/caches/Appveyor.cmake` by doing:
```set(OMR_PORT_SOCKET_SUPPORT ON CACHE BOOL "")```
> This also means that we will have to disable socket support in windows CI builds.

caohaley commented:
@babsingh @rwy0717 I believe it is ready for 2nd round of review. Also Robert, could you run the additional tests?

rwy0717 commented:
@genie-omr build all

caohaley commented on port/common/omrsockptb.c:
```diff
@@ -0,0 +1,138 @@
+ *
+ * @return a pointer to the per thread buffer on success, otherwise return NULL.
+ */
+OMRSocketPTB *
```
Sounds good, I will move it to this commit. That makes sense~

caohaley commented on cmake/caches/Appveyor.cmake:
```diff
@@ -33,4 +33,6 @@ set(OMR_THR_CUSTOM_SPIN_OPTIONS ON CACHE BOOL "")
 set(OMR_THR_THREE_TIER_LOCKING ON CACHE BOOL "")
 
-set(OMR_PORT_SOCKET_SUPPORT ON CACHE BOOL "")
+if(NOT OMR_HOST_OS STREQUAL "win")
```
@babsingh Sounds good, I will move them to a standalone commit. Also sorry just to clarify, for disabling tests on platforms, do you mean disabling the omrsockTest.cpp for windows platforms but we will still compile omrsock.c for windows?

babsingh replied:
correct. you need to provide a stub API implementation for `win32` which compiles.

caohaley commented on port/common/omrsockptb.h:
```diff
@@ -0,0 +1,41 @@
+#include "omrsock.h"
+#include "thread_api.h"
+
+OMRSocketPTB *omrsock_ptb_get(struct OMRPortLibrary *portLibrary);
```
Yes, I think we can. I will move it here for now. If there is anything that requires separate definition for OMRSocketPTB for different platforms, we can move them back into omrsock.h.

caohaley commented:
@babsingh @rwy0717 I believe it is ready for another round of review. After it finishes testing on here. It runs on my computer.

babsingh commented on fvtest/porttest/CMakeLists.txt:
```diff
@@ -42,7 +42,7 @@ if(OMR_OPT_CUDA)
 endif()
 
-if(OMR_PORT_SOCKET_SUPPORT)
+if((OMR_PORT_SOCKET_SUPPORT) AND (NOT OMR_HOST_OS STREQUAL "win"))
```
It should have a TODO comment indicating to remove `AND (NOT OMR_HOST_OS STREQUAL "win")` from the boolean expression after the OMRSock API is implemented on Windows.

babsingh commented on fvtest/porttest/makefile:
```diff
@@ -54,7 +54,9 @@ OBJECTS := \
 
 ifeq (1,$(OMR_PORT_SOCKET_SUPPORT))
-  OBJECTS += omrsockTest
+  ifneq (win,$(OMR_HOST_OS))
```
It should have a TODO comment indicating to remove `ifneq (win,$(OMR_HOST_OS))` after the OMRSock API is implemented on Windows.

babsingh commented on fvtest/porttest/makefile:
```diff
@@ -54,7 +54,9 @@ OBJECTS := \
   main_function
```
Copyright:
```
# Copyright (c) 2015, 2020 IBM Corp. and others
```

babsingh commented on fvtest/porttest/omrsockTest.cpp:
```diff
@@ -23,6 +23,7 @@
 #if defined(OMR_PORT_SOCKET_SUPPORT)
```
Copyright:
```
 * Copyright (c) 2019, 2020 IBM Corp. and others
```

rwy0717 commented on port/common/omrsock.c:
```diff
@@ -369,4 +384,17 @@ omrsock_close(struct OMRPortLibrary *portLibrary, omrsock_socket_t *sock)
+ */
+int32_t
+omrsock_shutdown(struct OMRPortLibrary *portLibrary)
+{	
```
```suggestion
{
```

rwy0717 commented on port/win32/omrsock.c:
```diff
@@ -0,0 +1,149 @@
+
+int32_t
+omrsock_shutdown(struct OMRPortLibrary *portLibrary)
+{	
```
```suggestion
{
```

rwy0717 commented on port/common/omrsock.c:
```diff
@@ -32,6 +32,21 @@
+ */
+int32_t
+omrsock_startup(struct OMRPortLibrary *portLibrary)
+{	
```
```suggestion
{
```

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -0,0 +1,180 @@
+
+int32_t
+omrsock_startup(struct OMRPortLibrary *portLibrary)
+{	
```
```suggestion
{
```

caohaley commented:
@rwy0717 I fixed the suggestions that you suggested. Can we run the additional tests?

rwy0717 commented:
@genie-omr build all

caohaley commented:
@rwy0717 Could you please run the tests again? I think I found why the segmentation fault happened. 

When I ran gdb on the Seg Fault, it seems like it happened because in `omrsock_ptb_shutdown`, I call `omrthread_tls_free` as long as portLibrary->portGlobals isn't NULL, no matter if `socketTlsKey` was allocated yet. 

I also realized that I don't set the `socketTlsKey` back to 0 after calling `omrthread_tls_free` so I added that in as well.

Therefore, I changed the implementation of `omrsock_ptb_shutdown` from:
```
int32_t 
omrsock_ptb_shutdown(struct OMRPortLibrary *portLibrary)
{
	if (NULL != portLibrary->portGlobals) {
		/* Release the TLS key */
		if(0 != omrthread_tls_free(portLibrary->portGlobals->socketTlsKey)){
			return OMRPORT_ERROR_SOCK_PTB_FAILED;
		}
	}
	return 0;
}
```
to
```
int32_t 
omrsock_ptb_shutdown(struct OMRPortLibrary *portLibrary)
{
	if ((NULL != portLibrary->portGlobals) && (0 != portLibrary->portGlobals->socketTlsKey)) {
		/* Release the TLS key */
		if(0 != omrthread_tls_free(portLibrary->portGlobals->socketTlsKey)){
			return OMRPORT_ERROR_SOCK_PTB_FAILED;
		}
		portLibrary->portGlobals->socketTlsKey = 0;
	}
	return 0;
}
```

babsingh commented:
@genie-omr build all

rwy0717 commented on port/common/omrsockptb.c:
```diff
@@ -0,0 +1,144 @@
+{
+	if ((NULL != portLibrary->portGlobals)  && (0 != portLibrary->portGlobals->socketTlsKey)) {
+		/* Release the TLS key */
+		if(0 != omrthread_tls_free(portLibrary->portGlobals->socketTlsKey)){
```
```suggestion
		if (0 != omrthread_tls_free(portLibrary->portGlobals->socketTlsKey)) {
```

rwy0717 commented:
@genie-omr build all

rwy0717 commented:
@genie-omr build all

rwy0717 commented:
@genie-omr build all

caohaley commented:
@rwy0717 The checks has passed.
